### DISCUSSION TITLE
Is it possible to cover multi-tenant scenarios with Prisma?

### DISCUSSION THREAD

**Discussion by manuelpoelzl** [prisma/prisma#14579] [Q&A] [answered]
Hi, I would like to migrate from TypeORM to Prisma and so I have the following question, can you cover multi tenant systems with Prisma?

For example, my application is built to connect to a database based on the user name.

Let's say I have the username "john@msystems", then it will look in my backend to see which database is associated with this organization and connect to that.

With TypeORM this can be solved via the ConnectionOptions, that you can specify a database name on the fly, but with Prisma I have only seen ways with .env files. This would not be possible in this case, because I simply cannot define all DB's as DataSource.

So now I have these two questions:
1. is the scenario described above possible with Prisma?
2. is it possible with Prisma, for example, to establish a connection to another DB in a function on the server, i.e. to open a new connection that is needed in this function, and then to close it again in this function?

Thanks in advance and sorry for the long text 😀

**Comment by janpio** [comment, marked as answer]
> 1. is the scenario described above possible with Prisma?

What you describe is probably easily possible in your application via the `datasources` parameter to `PrismaClient`: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#datasources + https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#programmatically-override-a-datasource-url

> 2. is it possible with Prisma, for example, to establish a connection to another DB in a function on the server, i.e. to open a new connection that is needed in this function, and then to close it again in this function?

That is harder to answer. You can alway instantiate a new PrismaClient with another connection string value.

There is currently no structured multi tenancy support yet (that would help manage migrations and so on).

**Reply by manuelpoelzl** [reply]
Thanks for your quick response!
Great to hear, that this is possible with the datasources parameter!
I´ll definitely try that.

Thank you very much!

**Reply by manuelpoelzl** [reply]
Hi @janpio 
I've been playing around with Prisma for the last few days, creating my models in the schema file.

However, what I have noticed now is that I can't get my PrismaClient to generate if I don't define a DataSource in it. In my case, I can't specify one here because, as I said, I don't have a "fixed" database to connect to. Is there a way to work around this, or is it possible to configure this somehow?

**Reply by janpio** [reply]
I think you can still define the `datasource`, but use a dummy value for `url` that follows the conventions but would never work.

**Reply by manuelpoelzl** [reply]
That worked out great! 
I still have one question, I have seen that you can map models to other column names, I also know this from TypeORM. But is there a way in Prisma to map columns that have a relation?

My model looks like this:

```javascript
model User{
@@map("users")

ID Int @id @default(autoincrement()) @unique
Firstname String @map("firstname") @db.VarChar(45)
Lastname String @map("lastname") @db.VarChar(45)
Phone String @map("phone") @db.VarChar(60) @unique
EMail String @map("email") @db.VarChar(45) @unique
Username String @map("username") @db.VarChar(45) @unique
Password String @map("password") @db.VarChar(255)
Role Role @map("role") @relation(fields:[Role],references:[ID])
IsLocked Boolean @map("isLocked") @default(false)
Avatar String? @db.VarChar(36) @unique
}
```

However, i´m getting the following error for the role column:
```
Error parsing attribute "@map": The attribute `@map` cannot be used on relation fields.
```

I didn't find anything in Prisma's documentation that deals with mapping relationships, so I wonder if this is possible in general. In TypeORM it was done in the following way:

```typescript

    @OneToOne((type) => Role, (role) => role.ID)
    @JoinColumn({name: "role"})
    Role: Role;
```

**Reply by janpio** [reply]
The relationship itself is not related to any field/column. Only the `fields` behind them are. Your data model currently uses relation field itself, that will not work. You will need a `Role_ID` or something similar next to it of type `Int` (same as the field it is referencing).

**Reply by manuelpoelzl** [reply]
Thank you so much for your help and detailed answers. Now everything works without problems and I can finally switch to Prisma!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
